### ISSUE TITLE
Neuron PJRT StableHLO compile rewrites sort+slice sampling path to AwsNeuronTopK with full vocab size as k

### ISSUE THREAD

**Issue by hugomano** [aws-neuron/aws-neuron-sdk#1339]
### Describe the bug

We operate an inference stack that compiles model graphs directly to the AWS Neuron PJRT plugin from StablehloHLO and shardy MLIR dialects. The workload is not using PyTorch/XLA or JAX as the frontend. We generate StableHLO ourselves, pass it to `PJRT_Client_Compile` with program format `mlir`, and rely on libneuronxla to compile it.

On the Neuron 2.30 stack with libneuronxla 3.0, a sampling graph containing a descending `stablehlo.sort` over logits followed by a top-3 slice fails during compilation. The compiler appears to rewrite the sort/slice pattern into an internal Neuron top-k custom call, but the generated `AwsNeuronTopK` uses the full `vocab` dimension (`128256`) as `k` instead of the slice size (`3`).

The observed compiler diagnostic is:
```
[NCC_EVRF014] TopK custom operation does not support k values larger than 16384. Found k=128256.
%custom-call.72 = (bf16[64,128256]{1,0}, s32[64,128256]{1,0}) custom-call(%reshape.71), custom_call_target="AwsNeuronTopK", backend_config="128256"
```

The source StableHLO pattern asks for top 3:
```
%0 = stablehlo.iota dim = 0 : tensor<128256xi32>
%1 = stablehlo.broadcast_in_dim %0, dims = [1] : (tensor<128256xi32>) -> tensor<64x128256xi32>
%2:2 = "stablehlo.sort"(%arg0, %1) <{dimension = 1 : i64, is_stable = true}> ({
^bb0(%arg2: tensor<bf16>, %arg3: tensor<bf16>, %arg4: tensor<i32>, %arg5: tensor<i32>):
  %33 = stablehlo.compare GT, %arg2, %arg3, FLOAT : (tensor<bf16>, tensor<bf16>) -> tensor<i1>
  stablehlo.return %33 : tensor<i1>
}) : (tensor<64x128256xbf16>, tensor<64x128256xi32>) -> (tensor<64x128256xbf16>, tensor<64x128256xi32>)
%3 = stablehlo.slice %2#0 [0:64, 0:3] : (tensor<64x128256xbf16>) -> tensor<64x3xbf16>
%5 = stablehlo.slice %2#1 [0:64, 0:3] : (tensor<64x128256xi32>) -> tensor<64x3xi32>
```
`128256` is the vocab/sort dimension. The requested top-k size is represented by the slice limit `0:3`.

If the compiler rewrites the idiom to `AwsNeuronTopK`, use the slice size (3) as the top k, not the full sorted dimension.

### Model Name

Observed in LLM inference sampling for different models.. One concrete workload was a Llama 3 model. The issue is not model-weight-specific.

### Describe the workload type

We are running our own inference server/frontend on top of StableHLO and the Neuron PJRT plugin:

This is important because the issue is not currently a PyTorch/XLA or JAX issue. A JAX `lax.top_k` control lowers to `chlo.top_k`, and a JAX sort-based control emits a different float-sort comparator:
```
%neg = stablehlo.negate %arg0
%sorted:2 = "stablehlo.sort"(%neg, %indices) ...
  %cmp = stablehlo.compare LT, %canonical_lhs, %canonical_rhs, TOTALORDER
%top_neg = stablehlo.slice %sorted#0 [0:64, 0:3]
%top = stablehlo.negate %top_neg
```
Our failing frontend emits plain StableHLO sort/slice.

### Instance Type

inf2.8xlarge

### Release version

Neuron SDK stack: 2.30
neuronx-cc: 2.25.3371.0+f524f7f8
libneuronxla: 3.0.2891.0
aws-neuronx-runtime-lib: 2.32.31.0
aws-neuronx-collectives: 2.32.28.0

### Reproduction Steps

The source StableHLO sort/slice part is:
```
%sorted_values_and_indices:2 = "stablehlo.sort"(%logits, %indices)
  <{dimension = 1 : i64, is_stable = true}> ({
^bb0(%lhs: tensor<bf16>, %rhs: tensor<bf16>, %lhs_idx: tensor<i32>, %rhs_idx: tensor<i32>):
  %before = stablehlo.compare GT, %lhs, %rhs, FLOAT
  stablehlo.return %before : tensor<i1>
}) : (tensor<64x128256xbf16>, tensor<64x128256xi32>)
  -> (tensor<64x128256xbf16>, tensor<64x128256xi32>)

%top_values = stablehlo.slice %sorted_values_and_indices#0 [0:64, 0:3]
%top_indices = stablehlo.slice %sorted_values_and_indices#1 [0:64, 0:3]
```

### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Possible Solution

If the sort/slice to topK rewrite is intended for this StableHLO pattern, the rewrite should use the post-sort slice size along the sorted dimension as k

**Comment by aws-liuliiily** [comment]
Hi @hugomano ,
  
Thank you for the detailed report and analysis. We've confirmed this is a bug in the sort-to-TopK rewrite pass — the pass is incorrectly using the full sort dimension size as k rather than extracting it from the downstream slice extent.

Customer Impact: This affects compilation on inf2 and trn1 instances when a stablehlo.sort followed by a slice is present in the graph. Compilation fails with NCC_EVRF014 when the sorted dimension exceeds 16384.

Workaround: You can manually tile the operation to keep k within the 16384 limit.

Next Steps: We have a fix planned by @devesr-amzn for a future Neuron SDK release. We'll update this issue when the fix is available. 

We appreciate you filing this with such a clear reproduction and diagnosis — it helped us triage quickly!

**Comment by hugomano** [comment]
Hey @aws-liuliiily @devesr-amzn , I updated our stack to Neuron 2.31.0 and the new error I have for the stablehlo sort/slice above is :
```
error(pjrt): [PJRT_Client_Compile] [ERROR] [NCC_EVRF014] TopK custom operation does not support k values larger than 16384. Found k=128256. Reduce the k value to 16384 or less.
%custom-call.61 = (bf16[128,128256]{1,0}, s32[128,128256]{1,0}) custom-call(%reshape.60), custom_call_target="AwsNeuronTopK", metadata={op_name="custom-call"}, backend_config="128256"
```

**Comment by aws-zhanxun** [comment]
Hi @hugomano,

Thanks for the update and for confirming the behavior on 2.31.0. You're right that the fix has not shipped yet, it wasn't included in this release, so 2.31.0 still hits the same NCC_EVRF014 error on the sort-slice pattern. 

The fix is still planned and is being tracked on our side. It will be available in an upcoming release.

In the meantime, the workaround from above still applies: manually tiling the operation so that k stays within the 16384 limit should unblock compilation.

Thanks again for the clear reproduction.
